### ISSUE TITLE
`Naming/PredicateMethod` false positive when an "operation" returns `true`

### ISSUE THREAD

**Issue by tagliala** [rubocop/rubocop#14232]
The new `Naming/PredicateMethod` is returning a false positive for a method which is, maybe, not the best method ever but it is definitely not a predicate method

```rb
def set_admin!
  raise 'Nope' unless some_checks?

  admin_arel.first_or_create
  true
end
```

Actually, the `true` is just used to return a simpler object here. I'm not incline to remove it and return the internal admin object, but if you think it should work that way I would do the change

--------

## Expected behavior

No offenses, because the method returns a boolean, but is not a predicate. The boolean is the "result of the operation"

## Actual behavior

```
test.rb:1:5: C: Naming/PredicateMethod: Predicate method names should end with ?.
def set_admin!
```

## Steps to reproduce the problem

1. Create a file `test.rb`

```rb
def set_admin!
  raise 'Nope' unless some_checks?

  admin_arel.first_or_create
  true
end
```

2. `bundle exec rubocop --only Naming/PredicateMethod test.rb` 

## RuboCop version

```
$ be rubocop -V
1.76.0 (using Parser 3.3.8.0, rubocop-ast 1.45.0, analyzing as Ruby 2.7, running on ruby 3.4.3) [arm64-darwin24]
```

From main branch

**Comment by bbatsov** [comment]
> Actually, the true is just used to return a simpler object here. I'm not incline to remove it and return the internal admin object, but if you thinl it should work that way I would do the change

I'm not sure there's much we can do in the cop here, as it behaves as expected. In your case I'd probably use different return value. (perhaps similar to how built-in `something!` methods operate)

**Comment by koic** [comment]
@bbatsov I think it's reasonable to classify this as a bug. For example, Active Record's `save!` method returns `true` when there are no issues, and raises an exception when there is a problem.
Perhaps, it would be appropriate to ignore methods like bang method (e.g, `do_something!`) that use `raise` in the method body.

**Comment by tagliala** [comment]
> For example, Active Record's save! method returns true when there are no issues, and raises an exception when there is a problem.

Thanks, I forgot if and where we took inspiration for that method, but I found an AR internal which is flagging an issue like this one, and does not have a raise

https://github.com/rails/rails/blob/b12f58620610f8247f897e7f396bb715bae93198/activerecord/lib/active_record/database_configurations/database_config.rb#L29-L33

It is a bang method indeed

**Comment by bbatsov** [comment]
@koic Fair enough. I guess we can add some special handling for bang methods, although I still find such logic a bit weird. (e.g. they could also return a symbol or the object in case of success, but I guess it is how it is)

**Comment by tagliala** [comment]
To provide more context, I ran this cop on a much larger project and it actually spot a method that should be a predicate instead

```
Offenses:

app/models/x.rb:62:7: C: Naming/PredicateMethod: Predicate method names should end with ?.
  def start!
      ^^^^^^
app/models/xx.rb:48:7: C: Naming/PredicateMethod: Predicate method names should end with ?.
  def start!(user_uid)
      ^^^^^^
app/models/x/x.rb:9:9: C: Naming/PredicateMethod: Predicate method names should end with ?.
    def upload_to_service
        ^^^^^^^^^^^^^^
app/models/xx/xx.rb:59:9: C: Naming/PredicateMethod: Predicate method names should end with ?.
    def reassign!
        ^^^^^^^^^
app/models/xxx/x.rb:18:11: C: Naming/PredicateMethod: Predicate method names should end with ?.
      def set_admin!
          ^^^^^^^^^^
app/models/xxx/x.rb:25:11: C: Naming/PredicateMethod: Predicate method names should end with ?.
      def unset_admin!
          ^^^^^^^^^^^^
app/models/xxx.rb:92:7: C: Naming/PredicateMethod: Predicate method names should end with ?.
  def cancel!
      ^^^^^^^
app/services/x/x.rb:58:9: C: Naming/PredicateMethod: Predicate method names should end with ?.
    def check_parameters!(step)
        ^^^^^^^^^^

2475 files inspected, 8 offenses detected
```

The following is actually a `bang` method, but we forgot the `!`, so this cop helped to identify an issue

```
app/models/x/x.rb:9:9: C: Naming/PredicateMethod: Predicate method names should end with ?.
    def upload_to_service
        ^^^^^^^^^^^^^^
```

**Comment by bbatsov** [comment]
Got it. Well, I think then this can be some configuration option that we can set to `true` for Rails projects.

**Comment by tagliala** [comment]
Thanks.

I think this is a useful cop, and I would not disable it on my Rails project. However, I'm asking if we can have some sort of configuration like `AllowedPatterns` like we have in example for `Metrics/MethodLength`, so we can exclude methods ending with `!`. This will also help to identify what is supposed to be a bang method but it isn't

**Comment by bbatsov** [comment]
Yeah, that's also a good idea. I'm guessing ideally we should do both:

- add some options like `AllowBangMethods`
- AllowedPatherns or AllowedMethods

All of those are pretty easy to implement.

**Comment by tagliala** [comment]
> AllowedMethods

This is already present and configured to allow `call` by default

**Comment by bbatsov** [comment]
Ah, OK. Seems there's even less work to do then. 😀
